onnx_tf/handlers/backend/tfidf_vectorizer.py

```diff
@@ -0,0 +1,123 @@
+    max_allowable_skip = np.minimum(max_skip,
+                                    int((int(x.shape[0]) - 1) / (n - 1) - 1))
+    for skip in range(max_allowable_skip + 1):
+      # For each skip calculate the ngram counts
+      result += cls._calc_ngram_skip(x, pool, n, skip)
+
+    return result
+
+  @classmethod
+  def _calc_ngram_skip(cls, x, pool, n, skip=0):
+    # This method calculates ngram counts for specific n and skip
+
+    # Make pool into an array of ngrams
+    pool = np.reshape(pool, (int(len(pool) / n), n))
+
+    # Make input as an array of ngrams
+    new_x = cls._prepare_ngrams(x, n, skip)
+
+    # Loop through the ngram targets in the pool
+    for i in range(len(pool)):
+      ngram_count = tf.map_fn(lambda in_x: tf.where(tf.reduce_all(tf.equal(in_x, tf.constant(pool[i], dtype=new_x.dtype))), tf.constant([1]), tf.constant([0])), new_x, dtype=tf.int32)
+      ngram_count = tf.math.count_nonzero(ngram_count, dtype=tf.int32)
+      ngram_count = tf.reshape(ngram_count, [1])
+      # Building up the result
+      if i == 0:
+        result = ngram_count
+      else:
+        result = tf.concat([result, ngram_count], 0)
+
+    return result
+
+  @classmethod
+  def version_9(cls, node, **kwargs):
+    input_tensor = kwargs["tensor_dict"][node.inputs[0]]
+    mode = node.attrs.get("mode")
+    max_skip_count = node.attrs.get("max_skip_count")
+    min_gram_len = node.attrs.get("min_gram_length")
+    max_gram_len = node.attrs.get("max_gram_length")
+    ngram_counts = node.attrs.get("ngram_counts")
+    ngram_indexes = node.attrs.get("ngram_indexes")
+    pool_int64s = node.attrs.get("pool_int64s")
+    pool_strings = node.attrs.get("pool_strings")
+    weights = node.attrs.get("weights", np.ones(len(ngram_indexes)))
+
+    def process_ngram(input_t):
+      # This is the main method that processes one batch of inputs
+      size = len(ngram_indexes)
+      new_ngram_counts = np.append(ngram_counts, size)
+      result_ngram = np.zeros(size)
+      for i in range(len(new_ngram_counts) - 1):
+        gram_len = i + 1
+        count = new_ngram_counts[i + 1] - new_ngram_counts[i]
+        total_len = count * gram_len
+        if gram_len >= min_gram_len and gram_len <= max_gram_len:
+          idx = ngram_indexes[new_ngram_counts[i]:new_ngram_counts[i + 1]]
+          process_pool = pool_int64s[
+              new_ngram_counts[i]:new_ngram_counts[i] +
+              total_len] if pool_int64s is not None else pool_strings[
+                  new_ngram_counts[i]:new_ngram_counts[i] + total_len]
+          result = cls._calc_ngram(input_t, process_pool, gram_len,
+                                   max_skip_count)
+          idx = tf.constant(idx, shape=[len(idx), 1])
+          result_ngram = result_ngram + tf.scatter_nd(idx, result, [size])
+      return result_ngram
+
+    # The input can be either 1d or 2d. Need to loop through
+    # each element for 2d inputs
+    n = len(input_tensor.shape)
+    final_out = [
```

Comment:
I appears to me that you have `process_ngram` organized as a function because you want the algorithm to work on two scenarios (1d, 2d). It may help with the clarity of the algorithm to have slightly more complicated pre/post-processing steps between which inputs are guaranteed to be 2d. This way, we reduce the complexity of the main body of the algorithm.

Reply:
Added comments to process_ngram to indicate general purpose of processing a row of inputs.